File: stackunderflow/etl/normalize/cursor_agent.py
"""Cursor Agent normalizer.

Cursor Agent persists turns as either marker-based plaintext
transcripts (legacy ``.txt``) or per-turn JSONL bubbles. **Neither
shape carries token counts.** Per the codeburn catalog, every Cursor
Agent record is estimated from text length / 4.

Therefore the normalizer's policy is unconditional: assistant rows
estimate ``input = len(content_text) // 4``, ``output = 0``, stamp
``cost_source='estimated'``. Even when an upstream future adds explicit
tokens we still want the estimated flag — Cursor Agent never reports
billing-grade counts.

Model id is forwarded from whatever the adapter resolved (the adapter
queries the ``conversation_summaries`` SQLite table for it); when that
fallback misses we use ``cursor-agent-auto``.
"""

from __future__ import annotations

import json
from collections.abc import Iterable

from .base import COST_SOURCE_ESTIMATED, Normalizer

_DEFAULT_MODEL = "cursor-agent-auto"
_RAW_EXTRAS_FIELDS = ("conversationId", "transcriptType", "toolCalls")


class CursorAgentNormalizer(Normalizer):
    provider_name = "cursor-agent"  # must equal the adapter's provider string

    def normalize(self, msg_row: dict) -> Iterable[dict]:
        role = str(msg_row.get("role") or "")
        if role != "assistant":
            return

        text = str(msg_row.get("content_text") or "")
        # Prefer the explicit estimate when an adapter pre-computed one
        # onto the input column; fall back to text//4 estimation here.
        input_tokens = int(msg_row.get("input_tokens") or 0)
        output_tokens = int(msg_row.get("output_tokens") or 0)
        if input_tokens == 0 and output_tokens == 0:
            if not text:
                return
            input_tokens = max(len(text) // 4, 0)

        if input_tokens == 0 and output_tokens == 0:
            return

        model = str(msg_row.get("model") or "") or _DEFAULT_MODEL

        raw_extras = _extras_from_raw_json(msg_row.get("raw_json"))

        yield self._build_event(
            msg_row,
            input_tokens=input_tokens,
            output_tokens=output_tokens,
            cache_read_tokens=0,
            cache_create_tokens=0,
            cost_source=COST_SOURCE_ESTIMATED,
            model=model,
            raw_extras=raw_extras,
        )


def _extras_from_raw_json(raw_json: object) -> dict | None:
    payload = _safe_load_raw(raw_json)
    if not isinstance(payload, dict):
        return None
    out: dict = {}
    for key in _RAW_EXTRAS_FIELDS:
        val = payload.get(key)
        if val is not None and val != "":
            out[key] = val
    return out or None


def _safe_load_raw(raw_json: object) -> object | None:
    if isinstance(raw_json, dict):
        return raw_json
    if not isinstance(raw_json, str | bytes | bytearray):
        return None
    try:
        if isinstance(raw_json, bytes | bytearray):
            return json.loads(raw_json.decode("utf-8", errors="replace"))
        return json.loads(raw_json)
    except (json.JSONDecodeError, ValueError, UnicodeDecodeError):
        return None

File: rust/crates/stax-etl/src/normalize/cursor_agent.rs
//! Cursor Agent — port of `etl/normalize/cursor_agent.py`.
//!
//! Neither on-disk shape (legacy marker plaintext, per-turn JSONL bubbles)
//! carries token counts, so the policy is unconditional: assistant rows
//! estimate `len(content_text) // 4` onto **input**, output stays 0, and the
//! stamp is always `estimated` — even when an adapter pre-computed counts,
//! because Cursor Agent never reports billing-grade numbers.
//!
//! **The registry key is `"cursor-agent"`, with the hyphen** — it must equal
//! the adapter's provider string. The hand-written registry block that shipped
//! this under the wrong key for months is the reason the Python registry became
//! self-discovering; the key is pinned by a test here and by the registry's own
//! parity test rather than left to a comment.

use super::base::{CostSource, EventSpec, NormalizeContext, Normalizer, UsageEvent};
use super::row::{MsgRow, PyRaise, int_or_zero, str_or_empty};
use super::text::{estimate_from_text, extras_from_payload};

const DEFAULT_MODEL: &str = "cursor-agent-auto";
const RAW_EXTRAS_FIELDS: [&str; 3] = ["conversationId", "transcriptType", "toolCalls"];

/// The `cursor-agent` normalizer.
#[derive(Debug, Clone, Copy, Default)]
pub struct CursorAgentNormalizer;

impl Normalizer for CursorAgentNormalizer {
    fn provider_name(&self) -> &'static str {
        "cursor-agent"
    }

    fn normalize(&self, ctx: &NormalizeContext, row: &MsgRow) -> Result<Vec<UsageEvent>, PyRaise> {
        if str_or_empty(row, "role") != "assistant" {
            return Ok(Vec::new());
        }

        let text = str_or_empty(row, "content_text");
        let mut input_tokens = int_or_zero(row, "input_tokens")?;
        let output_tokens = int_or_zero(row, "output_tokens")?;
        if input_tokens == 0 && output_tokens == 0 {
            if text.is_empty() {
                return Ok(Vec::new());
            }
            input_tokens = estimate_from_text(&text);
        }

        if input_tokens == 0 && output_tokens == 0 {
            return Ok(Vec::new());
        }

        let model = match str_or_empty(row, "model") {
            empty if empty.is_empty() => DEFAULT_MODEL.to_string(),
            model => model,
        };

        Ok(vec![
            self.build_event(
                ctx,
                row,
                EventSpec::new(input_tokens, output_tokens, 0, 0, CostSource::Estimated)
                    .model(model)
                    .raw_extras(extras_from_payload(row.get("raw_json"), &RAW_EXTRAS_FIELDS)),
            ),
        ])
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::normalize::test_support::{assistant_row, ctx};
    use stax_core::queries::pyjson::Value as PyValue;

    fn agent_row() -> MsgRow {
        assistant_row("cursor-agent", "claude-sonnet-4-5-20250929")
    }

    #[test]
    fn the_key_carries_the_hyphen_the_adapter_writes() {
        assert_eq!(CursorAgentNormalizer.provider_name(), "cursor-agent");
    }

    #[test]
    fn estimation_is_unconditional_even_for_a_rate_card_model() {
        let row = agent_row().with("content_text", PyValue::Str("x".repeat(44)));
        let events = CursorAgentNormalizer.normalize(&ctx(), &row).unwrap();
        assert_eq!((events[0].input_tokens, events[0].output_tokens), (11, 0));
        assert_eq!(events[0].cost_source, CostSource::Estimated);
    }

    #[test]
    fn pre_computed_counts_are_kept_but_still_stamped_estimated() {
        let row = agent_row()
            .with("input_tokens", PyValue::Int(300))
            .with("output_tokens", PyValue::Int(20))
            .with("content_text", PyValue::Str("ignored".into()));
        let events = CursorAgentNormalizer.normalize(&ctx(), &row).unwrap();
        assert_eq!((events[0].input_tokens, events[0].output_tokens), (300, 20));
        assert_eq!(events[0].cost_source, CostSource::Estimated);
    }

    #[test]
    fn an_empty_turn_yields_nothing() {
        assert!(
            CursorAgentNormalizer
                .normalize(&ctx(), &agent_row())
                .unwrap()
                .is_empty()
        );
        let tiny = agent_row().with("content_text", PyValue::Str("abc".into()));
        assert!(
            CursorAgentNormalizer
                .normalize(&ctx(), &tiny)
                .unwrap()
                .is_empty()
        );
    }
}
